// @xboxreplay/xboxlive-api/dist/errors.js
"use strict";
const errors_1 = require("@xboxreplay/errors");
const HTTPStatusCodes = require("./http-status-codes");
const errors = {
    internal: (message = 'Something went wrong...', statusCode = HTTPStatusCodes.INTERNAL_SERVER_ERROR) => new errors_1.XboxReplayError(message, {
        statusCode,
        reason: 'INTERNAL_ERROR'
    }),
    forbidden: (message = 'Forbidden', statusCode = HTTPStatusCodes.FORBIDDEN) => new errors_1.XboxReplayError(message, {
        statusCode,
        reason: 'FORBIDDEN'
    }),
    unauthorized: (message = 'Unauthorized', statusCode = HTTPStatusCodes.UNAUTHORIZED) => new errors_1.XboxReplayError(message, {
        statusCode,
        reason: 'UNAUTHORIZED'
    }),
    badRequest: (message = 'Bad request', statusCode = HTTPStatusCodes.BAD_REQUEST) => new errors_1.XboxReplayError(message, {
        statusCode,
        reason: 'BAD_REQUEST'
    }),
    requestError: (message = 'Request error', statusCode = HTTPStatusCodes.BAD_REQUEST) => new errors_1.XboxReplayError(message, {
        statusCode,
        reason: 'REQUEST_ERROR'
    })
};
module.exports = errors;
