// @xboxreplay/xboxlive-auth/dist/errors.js
"use strict";
const errors_1 = require("@xboxreplay/errors");
const HTTPStatusCodes = require("./http-status-codes");
const GitHubLinks = require("./github-links");
const errors = {
    internal: (message = 'Something went wrong...') => new errors_1.XboxReplayError(message, {
        statusCode: HTTPStatusCodes.INTERNAL_SERVER_ERROR,
        reason: 'INTERNAL_ERROR'
    }),
    matchError: (message = 'Match error') => new errors_1.XboxReplayError(message, {
        statusCode: HTTPStatusCodes.BAD_REQUEST,
        reason: 'MATCH_ERROR'
    }),
    invalidCredentials: (message = `Invalid credentials or 2FA enabled, please refer to ${GitHubLinks.twoFactorAuthenticationError}`) => new errors_1.XboxReplayError(message, {
        statusCode: HTTPStatusCodes.UNAUTHORIZED,
        reason: 'INVALID_CREDENTIALS'
    }),
    unauthorizedActivity: (message = `Activity confirmation required, please refer to ${GitHubLinks.unauthorizedActivityError}`) => new errors_1.XboxReplayError(message, {
        statusCode: HTTPStatusCodes.UNAUTHORIZED,
        reason: 'UNAUTHORIZED_ACTIVITY'
    }),
    exchangeFailure: (message = 'Exchange failure') => new errors_1.XboxReplayError(message, {
        statusCode: HTTPStatusCodes.BAD_REQUEST,
        reason: 'EXCHANGE_FAILURE'
    })
};
module.exports = errors;
